private/bufpkg/bufplugin/bufplugindocker/docker.go

```diff
@@ -0,0 +1,213 @@
+		return buildkitSession.Run(errGroupCtx, func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
+			return d.cli.DialHijack(ctx, "/session", proto, meta)
+		})
+	})
+
+	buildID := stringid.GenerateRandomID()
+	imageName := pluginConfig.Name.IdentityString() + ":" + buildID
+	eg.Go(func() (retErr error) {
+		defer func() {
+			if err := buildkitSession.Close(); err != nil {
+				retErr = multierr.Append(retErr, err)
+			}
+		}()
+
+		response, err := d.cli.ImageBuild(ctx, dockerContext, types.ImageBuildOptions{
+			Tags:     []string{imageName},
+			Platform: "linux/amd64",
+			Labels: map[string]string{
+				"build.buf.plugins.config.remote": pluginConfig.Name.Remote(),
+				"build.buf.plugins.config.owner":  pluginConfig.Name.Owner(),
+				"build.buf.plugins.config.name":   pluginConfig.Name.Plugin(),
+			},
+			Version:   types.BuilderBuildKit, // DOCKER_BUILDKIT=1
+			SessionID: buildkitSession.ID(),
+		})
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := response.Body.Close(); err != nil {
+				retErr = multierr.Append(retErr, err)
+			}
+		}()
+		scanner := bufio.NewScanner(response.Body)
+		for scanner.Scan() {
+			d.logger.Debug(scanner.Text())
+		}
+		if err := scanner.Err(); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	imageInfo, _, err := d.cli.ImageInspectWithRaw(ctx, imageName)
+	if err != nil {
+		return nil, err
+	}
+	return &BuildResponse{
+		Image:  imageName,
+		Digest: imageInfo.ID,
+	}, nil
+}
+
+func (d *dockerAPIClient) Push(ctx context.Context, image string, auth *RegistryAuthConfig) (response *PushResponse, retErr error) {
+	registryAuth, err := auth.ToHeader()
+	if err != nil {
+		return nil, err
+	}
+	pushReader, err := d.cli.ImagePush(ctx, image, types.ImagePushOptions{
+		RegistryAuth: registryAuth,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, pushReader.Close())
+	}()
+	pushScanner := bufio.NewScanner(pushReader)
+	for pushScanner.Scan() {
+		d.logger.Debug(pushScanner.Text())
+	}
+	if err := pushScanner.Err(); err != nil {
+		return nil, err
+	}
+	return &PushResponse{}, nil
+}
+
+func (d *dockerAPIClient) Delete(ctx context.Context, image string) (*DeleteResponse, error) {
+	_, err := d.cli.ImageRemove(ctx, image, types.ImageRemoveOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return &DeleteResponse{}, nil
+}
+
+func (d *dockerAPIClient) Close() error {
+	return d.cli.Close()
+}
+
+// NewClient creates a new Client to use to build Docker plugins.
+func NewClient(logger *zap.Logger) (Client, error) {
+	if logger == nil {
+		return nil, errors.New("logger required")
+	}
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, err
+	}
+	dockerClient := &dockerAPIClient{cli: cli, logger: logger}
+	return dockerClient, nil
```

Comment:
nit: I would just return the instance without a variable declaration.
```suggestion
	return &dockerAPIClient{
	    cli: cli,
	    logger: logger,
	}, nil
```